src/scipp/format/formatter.py

```diff
@@ -1,40 +1,130 @@
-def format_variable(data, spec):
-    """
-    String formats the Variable according to the provided specification.
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+# @author Gregory Tucker, Jan-Lukas Wynen
+
+from typing import Any, List
+
+import numpy as np
+
+from ..core.cpp_classes import DType, Unit, Variable
+from ._parse import FormatSpec, FormatType, parse
+
+
+def format_variable(self, format_spec: str) -> str:
+    """String formats the Variable according to the provided specification.
 
     Parameters
     ----------
-    data
-        A scalar or array-like scipp Variable object
-    spec
-        Format specification; only 'c' for Compact error-reporting supported at present
+    format_spec:
+        Format specification;
+        only 'c' for Compact error-reporting supported at present.
 
     Returns
     -------
-    The formatted string
+    :
+        The formatted string.
     """
-    from numpy import array
 
-    from ..core.cpp_classes import Unit
-    dtype = str(data.dtype)
-    if not any([x in dtype for x in ('float', 'int')]) or spec is None or len(spec) < 1:
-        return data.__repr__()
-    compact = spec[-1] == 'c'
+    spec = parse(format_spec, Variable)
+    return _VARIABLE_FORMATTERS[spec.format_type](self, spec)
+
+
+def _format_sizes(data: Variable) -> str:
+    return '(' + ', '.join(f'{dim}: {size}' for dim, size in data.sizes.items()) + ')'
+
+
+def _format_unit(data: Variable) -> str:
+    if data.unit is None:
+        return '<no unit>'
+    return f'[{data.unit}]'
+
+
+def _format_element(elem: Any, spec: str):
+    if spec:
+        return f'{elem:{spec}}'
+    if isinstance(elem, (int, float)):
+        # Replicate behavior of C++ formatter.
+        return f'{elem:g}'
+    if isinstance(elem, str):
+        return f'"{elem}"'
+    return f'{elem}'
+
+
+def _as_flat_array(data):
+    if isinstance(data, np.ndarray):
+        return data.flat
+    if 'ElementArray' in repr(type(data)):
+        return data
+    return np.array([data])
+
 
-    val = data.values if data.shape else array((data.value, ))
-    var = data.variances if data.shape else array((data.variance, ))
-    unt = "" if data.unit == Unit('dimensionless') else f" {data.unit}"
+def _format_array_flat(data, length: int, dtype: DType, spec: str) -> str:
+    if dtype in (DType.Variable, DType.DataArray, DType.Dataset, DType.VariableView,
+                 DType.DataArrayView, DType.DatasetView):
```

Comment:
`DataGroup` currently leads to `dtype=PyObject`, do we want/need special handling?

Reply:
We need some special handling. Right now this is not good:
```python
v = sc.scalar(sc.DataGroup({'a': 2.1}))
str(v)  # <scipp.Variable> ()   PyObject        <no unit>  [DataGroup(sizes={}, keys=['a'])]
f'{v}'  # <scipp.Variable> ()   PyObject        <no unit>  [['a']]
```

Reply:
Since we don't have custom formatting for data group (and that would be difficult to do anyway), I changed it to fall back to using `str(data)`.